### PR TITLE
Fix for UT failure at protects kube objects

### PR DESCRIPTION
**Pull request by pruthvitd**
Fix for "protects kube objects" UT failure.
Following are the few of the reasons why we see UT failures:
1. There is inconsistent upload and download of the object, as part of test these objects are compared and failed to match.
The sequence of the steps in the test are as follows:

> 1. Upload is done by VRG controller.
> 2. Download is done by ProtectedVolumeReplicationGroupList(PVRGL) and the the object is updated within its CR Status.
> 3. From the test the object is downloaded from fake S3 again and compared against the PVRGL status. 

In the above steps the problem is in between the step#1 and step#2, because even before VRG object is uploaded to S3 bucket PVRGL downloads it. The downloaded objects are compared against PVRGL status and we notice a mismatch.

Fix: This has been fixed by waiting until condition ClusterDataProtected.Reason=Uploaded to create PVRGL object.

2. Same fake S3 bucket (which is a map of key and an k8s object) used in every test context without cleanup of the map. The stale data from the previous tests caused inconsistent behavior in the trailing tests.

3. Within the VRG test, vrg object status is updated which conflicts with the updates done on the object by the controller leading to errors like :
```
 Operation cannot be fulfilled on volumereplicationgroups.ramendr.openshift.io "vrg-lgeec": the object has been modified; please apply your changes to the latest version and try again
```
4. VRG controller keeps uploading the VRG object to S3 bucket with its updated "lastUpdateTime" which is being compared against the same VRG object downloaded by ProtectedVolumeReplicationGroupList controller. This leads to object mismatch due to the difference in timestamp.

Image

Tested locally and it works.
```
/root/DR/VM_DR_Within_NS/defect-fix/fence-off/blocker-fix/latest/latest/ramen/bin/controller-gen object:headerFile="hack/boilerplate.go.txt" paths="./..."
/root/DR/VM_DR_Within_NS/defect-fix/fence-off/blocker-fix/latest/latest/ramen/bin/controller-gen rbac:roleName=operator-role crd:generateEmbeddedObjectMeta=true webhook paths="./..." output:crd:artifacts:config=config/crd/bases
hack/install-setup-envtest.sh
go test ./... -coverprofile cover.out
	github.com/ramendr/ramen/hack/fakes		coverage: 0.0% of statements
	github.com/ramendr/ramen/cmd		coverage: 0.0% of statements
?   	github.com/ramendr/ramen/internal/controller/core	[no test files]
	github.com/ramendr/ramen/internal/controller/kubeobjects/velero		coverage: 0.0% of statements
	github.com/ramendr/ramen/internal/controller/testutils		coverage: 0.0% of statements
	github.com/ramendr/ramen/internal/controller/argocd		coverage: 0.0% of statements
ok  	github.com/ramendr/ramen/internal/controller	198.013s	coverage: 64.4% of statements
ok  	github.com/ramendr/ramen/internal/controller/cel	27.750s	coverage: [no statements]
ok  	github.com/ramendr/ramen/internal/controller/cephfscg	26.694s	coverage: 64.6% of statements
ok  	github.com/ramendr/ramen/internal/controller/hooks	0.493s	coverage: 43.0% of statements
ok  	github.com/ramendr/ramen/internal/controller/kubeobjects	0.325s	coverage: 33.3% of statements
ok  	github.com/ramendr/ramen/internal/controller/util	25.274s	coverage: 25.1% of statements
ok  	github.com/ramendr/ramen/internal/controller/volsync	41.292s	coverage: 45.6% of statements
```

Fixes: #2060

### CODE REVIEW

**Review by abhijeet219 — APPROVED** [review]

File: internal/controller/vrg_volrep_test.go [hunk 1]
```diff
@@ -953,6 +964,9 @@ var _ = Describe("VolumeReplicationGroupVolRepController", func() {
 		})
 		It("cleans up after testing", func() {
 			vrgCreateVGRTestCase.cleanupProtected()
+			Eventually(func() error {
+				return (*vrgObjectStorer).DeleteObjectsWithKeyPrefix(vrgCreateVGRTestCase.s3KeyPrefix())
```
Comment (nirs) [inline internal/controller/vrg_volrep_test.go:968]: This is the fake store, right? why do we need to retry the delete multiple times? How can it fail?

I see this is repeated many times - can we use After helper to run in after all tests that use s3 store?

If we cannot use the After helper but we need to do this at the end of many tests, and this should have retries with a timeout, we need to add a helper that does this work, and call it from the test, so we don't need to duplicate the boilerplate. With a helper you change the way you clean up by changing one function instead of changing many functions.

In this case the entire block under:

```go
It("cleans up after testing" func() {
```

Looks like a cleanup function that can use the standard way to run cleanup after a test.

File: internal/controller/vrg_volrep_test.go [hunk 2]
```diff
@@ -2348,6 +2395,37 @@ func (v *vrgTest) verifyVRGStatusExpectation(expectedStatus bool, reason string)
 		"while waiting for VRG TRUE condition %s/%s", v.vrgName, v.namespace)
 }
 
+func (v *vrgTest) verifyVRGStatusExpectationForS3Upload(expectedStatus bool, reason string) {
```
Comment (nirs) [inline internal/controller/vrg_volrep_test.go:2398]: verifyVRGStatusForS3Upload?

It is not clear if this should be caller before the upload to check if we can upload or after the upload to check if the upload was correct. The name of the function should make this clear.

It the name cannot explain the complicated assumptions, add a comment explaining when this should be called and why.

**Comment by pruthvitd** [comment]
> Please add a commit message with this info:
> 
>     * What it the issue being fixed
> 
>     * Link to the issue if we have one
> 
>     * What is the root cause
> 
>     * What is the fix
> 
>     * Alternatives considered
> 
> 
> Without this I don't have any context to review this change.

Hello @nirs Please find the details on this git issue: https://github.com/RamenDR/ramen/issues/2060 and also I have updated the PR description.

Thanks

File: internal/controller/vrg_volrep_test.go (hunk 1 above)
Comment (pruthvitd) [reply internal/controller/vrg_volrep_test.go:968]: cleanup is a generic method and here we need to delete the VRG object that is created within this context, as quick solution its deleted within this loop. Else the loop will have to be moved to cleanup() method.

File: internal/controller/vrg_volrep_test.go (hunk 2 above)
Comment (pruthvitd) [reply internal/controller/vrg_volrep_test.go:2398]: sure renaming to "ensureVRGIsUploadedToS3"

**Comment by nirs** [comment]
> Fix for "protects kube objects" UT failure. Following are the few of the reasons why we see UT failures:
> 
> 1. There is inconsistent upload and download of the object, as part of test these objects are compared and failed to match.
>    The sequence of the steps in the test are as follows:
> 
> > 1. Upload is done by VRG controller.
> > 2. Download is done by ProtectedVolumeReplicationGroupList(PVRGL) and the the object is updated within its CR Status.
> > 3. From the test the object is downloaded from fake S3 again and compared against the PVRGL status.
> 
> In the above steps the problem is in between the step#1 and step#2, because even before VRG object is uploaded to S3 bucket PVRGL downloads it. The downloaded objects are compared against PVRGL status and we notice a mismatch.

How PVRGL downloads the object before it was uploaded?

> Fix: This has been fixed by waiting until condition ClusterDataProtected.Reason=Uploaded to create PVRGL object.

This seems to cover the real issue - how a n object exists in s3 before we uploaded it.

> 2. Same fake S3 bucket (which is a map of key and an k8s object) used in every test context without cleanup of the map. The stale data from the previous tests caused inconsistent behavior in the trailing tests.

So the root cause it this:

1. Test 1 uploaded vrg to s3
2. Test 2 downloads the same vrg from s3
3. Test 2 verify vrg before it was uploaded by ramen

> 3. Within the VRG test, vrg object status is updated which conflicts with the updates done on the object by the controller leading to errors like :
> 
> ```
>  Operation cannot be fulfilled on volumereplicationgroups.ramendr.openshift.io "vrg-lgeec": the object has been modified; please apply your changes to the latest version and try again
> ```

This is a separate issue. We should always use retry on conflict helper when updating.

> 4. VRG controller keeps uploading the VRG object to S3 bucket with its updated "lastUpdateTime" which is being compared against the same VRG object downloaded by ProtectedVolumeReplicationGroupList controller. This leads to object mismatch due to the difference in timestamp.

So we have several issues:
- Tests do not clean up when they finish
- Using the same vrg name in multiple tests - test should be isolated when possible. There is no reason to use the same vrg name in different tests
- Unhandled conflicts when updating resources
